File: apps/api/audit-logs/repo.ts
import { and, count, desc, eq, gte, lte, SQL } from 'drizzle-orm'
import log from 'encore.dev/log'
import { AuditLogRepository } from '.'
import orm, { DrizzleDatabase } from '../database'
import {
	AuditLog,
	AuditLogDB,
	AuditLogParams,
	AuditLogQuery,
	auditLogs
} from '../schema/audit-logs'
import { handleDatabaseErr } from '../utils'

const DEFAULT_PAGE_SIZE = 20

class repo implements AuditLogRepository {
	constructor(private readonly db: DrizzleDatabase) {}

	create(params: AuditLogParams): Promise<AuditLogDB> {
		log.info('AuditLogRepo.create params: ', { params })
		return this.db
			.insert(auditLogs)
			.values(params)
			.returning()
			.then((rows) => rows[0])
			.catch(handleDatabaseErr)
	}

	async find(
		query: AuditLogQuery
	): Promise<{ data: AuditLog[]; total: number }> {
		const conditions: SQL[] = []
		if (query.resource !== undefined) {
			conditions.push(eq(auditLogs.resource, query.resource))
		}
		if (query.action !== undefined) {
			conditions.push(eq(auditLogs.action, query.action))
		}
		if (query.actorUserId !== undefined) {
			conditions.push(eq(auditLogs.actorUserId, query.actorUserId))
		}
		if (query.from !== undefined) {
			conditions.push(gte(auditLogs.createdAt, query.from))
		}
		if (query.to !== undefined) {
			conditions.push(lte(auditLogs.createdAt, query.to))
		}

		const where = conditions.length === 0 ? undefined : and(...conditions)

		const page = query.page && query.page > 0 ? query.page : 1
		const pageSize =
			query.pageSize && query.pageSize > 0
				? query.pageSize
				: DEFAULT_PAGE_SIZE

		const [data, totalResult] = await Promise.all([
			this.db.query.auditLogs.findMany({
				where,
				with: { actor: true },
				orderBy: desc(auditLogs.createdAt),
				limit: pageSize,
				offset: (page - 1) * pageSize
			}),
			this.db.select({ total: count() }).from(auditLogs).where(where)
		]).catch(handleDatabaseErr)

		return {
			data: data as unknown as AuditLog[],
			total: totalResult[0]?.total ?? 0
		}
	}
}

const auditLogRepo = new repo(orm)

export default auditLogRepo

File: apps/api-go/internal/auditlogs/repo.go
// Package auditlogs is the persistence layer for the audit_logs table.
// Rows are only ever written by the pubsub subscription in
// auditlogs/auditlogs.go, never directly by an API handler, so — like
// internal/materialassetevents — this repo only supports Create/Find, no
// Update/Delete.
package auditlogs

import (
	"context"
	"fmt"

	"encore.app/internal/entities"
	"github.com/pocketbase/dbx"
)

type Repository struct {
	db *dbx.DB
}

func NewRepository(db *dbx.DB) *Repository {
	return &Repository{db: db}
}

func (r *Repository) Create(ctx context.Context, row *entities.AuditLog) error {
	if err := r.db.WithContext(ctx).Model(row).Insert(); err != nil {
		return fmt.Errorf("audit_logs: insert: %w", err)
	}
	return nil
}

// Query narrows Find; zero values mean "no filter" for that field.
type Query struct {
	Resource    string
	Action      entities.AuditAction
	ActorUserID int64
	From        string
	To          string
	Page        int
	PageSize    int
}

const defaultPageSize = 20

func (r *Repository) Find(ctx context.Context, q Query) (rows []*entities.AuditLog, total int64, err error) {
	var conds []dbx.Expression
	if q.Resource != "" {
		conds = append(conds, dbx.HashExp{"resource": q.Resource})
	}
	if q.Action != "" {
		conds = append(conds, dbx.HashExp{"action": q.Action})
	}
	if q.ActorUserID != 0 {
		conds = append(conds, dbx.HashExp{"actorUserId": q.ActorUserID})
	}
	if q.From != "" {
		conds = append(conds, dbx.NewExp("createdAt >= {:from}", dbx.Params{"from": q.From}))
	}
	if q.To != "" {
		conds = append(conds, dbx.NewExp("createdAt <= {:to}", dbx.Params{"to": q.To}))
	}

	page := q.Page
	if page <= 0 {
		page = 1
	}
	pageSize := q.PageSize
	if pageSize <= 0 {
		pageSize = defaultPageSize
	}

	where := dbx.Expression(nil)
	if len(conds) > 0 {
		where = dbx.And(conds...)
	}

	countQ := r.db.WithContext(ctx).Select("COUNT(*)").From("audit_logs")
	if where != nil {
		countQ = countQ.Where(where)
	}
	if err := countQ.Row(&total); err != nil {
		return nil, 0, fmt.Errorf("audit_logs: count: %w", err)
	}

	dataQ := r.db.WithContext(ctx).Select("*").From("audit_logs").
		OrderBy("createdAt DESC").
		Limit(int64(pageSize)).
		Offset(int64((page - 1) * pageSize))
	if where != nil {
		dataQ = dataQ.Where(where)
	}
	if err := dataQ.All(&rows); err != nil {
		return nil, 0, fmt.Errorf("audit_logs: find: %w", err)
	}

	return rows, total, nil
}
